### ISSUE TITLE
Switch from libseccomp to BPF assembly

### ISSUE THREAD

**Issue by zenhack** [sandstorm-io/sandstorm#2989]
From <https://github.com/sandstorm-io/sandstorm/blob/5ee73e6022e8d3620d48e082a08a70a725795c9d/src/sandstorm/supervisor.c%2B%2B#L1145-L1147>:

```c++
// TODO(security): We should probably disallow everything except SOCK_STREAM and SOCK_DGRAM but
//   I don't totally get how to write such conditionals with libseccomp. We should really dump
//   libseccomp and write in BPF assembly, which is frankly much easier to understand.
```

I didn't see an open issue for this, so thought I'd add one. It's seems like most of the comments in that section refer to things that wouldn't be hard to deal with using raw BPF asm.

**Comment by kentonv** [comment]
Yeah, I've been working with BPF asm for Cloudflare Workers, and it's really quite nice.

**Comment by zenhack** [comment]
Yeah, I've played around with it and it *is* fairly nice.

What tools do you use to actually assemble it? There's an assembler
under tools/ in the linux source tree, but I don't know what's standard.

**Comment by kentonv** [comment]
Someone at Cloudflare extracted the tools so that they're independently buildable (before I joined, actually!): https://github.com/cloudflare/bpftools

I added some additional preprocessing so that I could #include C headers and evaluate constant expressions, but it's a pretty terrible hack, not sure if it's even worth releasing.

**Comment by zenhack** [comment]
A while back I fooled around with trying to do this to get a sense of what it might look like.
What I ended up doing was just piping the assmebly source through cpp, and #including the output of this program from the assembly:

https://gist.github.com/zenhack/9e7c27b00893fff116b168ae85056ff9

...which got run as part of the build. It didn't seem worth the effort to try to automate it any further.

**Comment by zenhack** [comment]
Just a note, I have some work on this on a branch, which also turns the filter into a whitelist (I started just working on the latter, but found it difficult enough to encode certain things that I decided to just go all-in on this). Still needs some debugging & testing though.
